### ISSUE TITLE
[ci] [docs] docs rendering failing with 'undefined label: 'metadata_routing'

### ISSUE THREAD

**Issue by jameslamb** [microsoft/LightGBM#5954]
## Description

The `check-docs` CI job is failing with this error that seems to be occurring while rendering the sphinx docs to HTML

> Warning, treated as error:
/home/runner/work/LightGBM/LightGBM/python-package/lightgbm/dask.py:docstring of lightgbm.dask.DaskLGBMClassifier:1:undefined label: 'metadata_routing'

([build link](https://github.com/microsoft/LightGBM/actions/runs/5424541292/jobs/9864047043?pr=5953))

## Reproducible example

I see that in multiple CI builds here, including the latest on `master` ([build link](https://github.com/microsoft/LightGBM/actions/runs/5423459086/jobs/9861417030)).

I'm able to reproduce this locally in a Python 3.10 environment on macOS, with the following:

```shell
mamba install --yes \
    breathe \
    doxygen \
    scikit-learn \
    sphinx \
    sphinx_rtd_theme \
    'scikit-learn>=1.3'
```

## Additional Notes

When downgrading from `scikit-learn` 1.3.0 to 1.2.2, this error goes away and the docs render correctly. I strongly suspect that it's related to changes in the most recent version of `scikit-learn`.

**Comment by jameslamb** [comment]
> strongly suspect that it's related to changes in the most recent version of `scikit-learn`

`scikit-learn` v1.3.0 was released 2 days ago ([link](https://github.com/scikit-learn/scikit-learn/releases/tag/1.3.0)), which coincides with when I first started seeing these CI failures.

I see the following in the release notes ([link](https://scikit-learn.org/stable/whats_new/v1.3.html#experimental-under-development))

> [Metadata routing](https://scikit-learn.org/stable/metadata_routing.html#metadata-routing)’s related base methods are included in this release. This feature is only available via the enable_metadata_routing feature flag which can be enabled using [sklearn.set_config](https://scikit-learn.org/stable/modules/generated/sklearn.set_config.html#sklearn.set_config) and [sklearn.config_context](https://scikit-learn.org/stable/modules/generated/sklearn.config_context.html#sklearn.config_context). For now this feature is mostly useful for third party developers to prepare their code base for metadata routing, and we strongly recommend that they also hide it behind the same feature flag, rather than having it enabled by default.

I'm still reading through this to understand what it means and why it'd show up as a breaking change in the documentation generation.

**Comment by jameslamb** [comment]
I think I see the problem.

All of `lightgbm`'s scikit-learn estimators inherit from `sklearn.base.BaseEstimator`. For example:

https://github.com/microsoft/LightGBM/blob/9f78cceee4911dd56f4635dfd36d4482363db5aa/python-package/lightgbm/compat.py#L73

https://github.com/microsoft/LightGBM/blob/9f78cceee4911dd56f4635dfd36d4482363db5aa/python-package/lightgbm/sklearn.py#L391

https://github.com/microsoft/LightGBM/blob/9f78cceee4911dd56f4635dfd36d4482363db5aa/python-package/lightgbm/sklearn.py#L1068

When we render this project's `sphinx` docs using `sphinx.ext.autodoc`, all of the public properties and methods (anything not beginning with a `_`) get a documentation entry. See, for example, https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.LGBMClassifier.html.

Prior to v1.3.0, `sklearn.base.BaseEstimator` had only two public methods and 0 public attributes... `get_params()` and `set_params()`. Both of those are overridden by LightGBM, and the docstrings are overridden as well.

https://github.com/microsoft/LightGBM/blob/9f78cceee4911dd56f4635dfd36d4482363db5aa/python-package/lightgbm/sklearn.py#L583-L584

https://github.com/microsoft/LightGBM/blob/9f78cceee4911dd56f4635dfd36d4482363db5aa/python-package/lightgbm/sklearn.py#L601-L602

As of v1.3.0, there's now a third public method, `BaseEstimator.get_metadata_routing`.

```python
from sklearn.base import BaseEstimator
[x for x in dir(BaseEstimator) if not x.startswith("_")]

# ['get_metadata_routing', 'get_params', 'set_params']
```

Since that isn't overridden by `lightgbm`, its docstring gets passed directly through to `lightgbm`'s docs-generation process. Unfortunately, that docstring contains a reference to another part of scikit-learn's documentation, which means results in the error reported at the top of this issue.

```python
from sklearn.base import BaseEstimator
BaseEstimator.get_metadata_routing.__doc__
```

shows:

```text
'Get metadata routing of this object.\n\n        Please check :ref:`User Guide <metadata_routing>` on how the routing\n        mechanism works.\n\n        Returns\n        -------\n        routing : MetadataRequest\n            A :class:`~utils.metadata_routing.MetadataRequest` encapsulating\n            routing information.\n        '
```

See that `<metadata_routing>`? That's a documentation link that's assuming you're building alongside `scikit-learn`'s own docs 🙃 

It's a reference to this scikit-learn-only `:ref:` https://github.com/scikit-learn/scikit-learn/blob/b88b53985d9ddf8cec60414934c55a5745e8b958/doc/user_guide.rst#L42

----

**Ok, so what do we do?**

I'm going to do the following right now to unblock development in this repo:

* put up a PR adding `get_metadata_routing()` to the public API of `LGBMModel`, with our own docstring and just transparently passing through to the method from `sklearn.base.BaseEstimator`

And then after that, I'll do the following:

* open a PR in `scikit-learn` removing that reference to the User Guide, so other projects like `xgboost` that implement scikit-learn-compatible estimators don't experience this same issue
* open a feature request here in LightGBM describing the need to add support for this new "metadata routing" concept to this project's `scikit-learn` estimators

**Comment by jameslamb** [comment]
cc @trivialfis @hcho3, you might see the same thing the next time `xgboost`'s docs build, since `xgboost` scikit-learn estimators also inherit from `sklearn.base`.

https://github.com/dmlc/xgboost/blob/f90771eec64cea7b1cb4e41a31000c8c54fcc6d8/python-package/xgboost/compat.py#L48-L50

Hopefully this write-up can save you some time.

**Comment by trivialfis** [comment]
Thank you for the detailed analysis @jameslamb ! I saw this PR: https://github.com/microsoft/LightGBM/pull/5956 by @thomasjpfan . XGBoost is using `intersphinx` as well for making references to related projects like numpy and pandas.

**Comment by Jacob-Stevens-Haas** [comment]
From another [repo](github.com/dynamicslab/pysindy) that inherits `BaseEstimator` and ran into the same issue, thanks for the research :tada: !

**Comment by jameslamb** [comment]
Sure, glad it helped you!
